// @geoblocks/mapfishprint/lib/BaseCustomizer.js
/**
 * The customizer allows to customize some transformations.
 * It also defines the print extent.
 */
export default class BaseCustomizer {
    printExtent;
    constructor(printExtent) {
        this.setPrintExtent(printExtent || [0, 0, Infinity, Infinity]);
    }
    getPrintExtent() {
        return this.printExtent;
    }
    setPrintExtent(printExtent) {
        this.printExtent = printExtent;
    }
    /**
     *
     * @param layerState
     * @return true to convert this layer, false to skip it
     */
    layerFilter(layerState) {
        return true;
    }
    /**
     * Decide to skip some geometries.
     * Useful to avoid sending features outside the print extend on the wire.
     * @param geometry
     * @return true to convert this feature, false to skip it
     */
    geometryFilter(geometry) {
        // FIXME: shouldn't we provide some reasonable defaults here?
        // For ex:
        // - define a buffer of X pixels and remove all points outside it;
        // - only keep lines / polygons that intersect it
        // Cf schm for some code.
        return true;
    }
    /**
     * Can be used to add / remove properties to features
     * @param layerState
     * @param feature converted feature
     */
    feature(layerState, feature) { }
    /**
     * Can be used to manipulate the line symbolizers
     * @param layerState
     * @param symbolizer
     * @param stroke
     */
    line(layerState, symbolizer, stroke) { }
    /**
     * Can be used to manipulate the image symbolizers
     * @param layerState
     * @param symbolizer
     * @param image
     */
    point(layerState, symbolizer, image) { }
    /**
     * Can be used to manipulate the text symbolizers
     * @param layerState
     * @param symbolizer
     * @param text
     */
    text(layerState, symbolizer, text) { }
    /**
     * Can be used to manipulate a converted WMTS layer
     * @param layerState
     * @param wmtsLayer
     * @param source
     */
    wmtsLayer(layerState, wmtsLayer, source) { }
    // FIXME: does it really makes sense?
    // Why isn't it done on an extended BaseEncoder instead?
    /**
     * Can be used to manipulate a converted WMS layer
     * @param layerState
     * @param wmsLayer
     * @param source
     */
    wmsLayer(layerState, wmsLayer, source) { }
}
