// @geoblocks/mapfishprint/lib/MapfishPrintBaseEncoder.js
/**
 * @module app.print.Service
 */
import { getWmtsMatrices } from "./mapfishprintUtils.js";
import olLayerTile from 'ol/layer/Tile';
import olSourceWMTS from 'ol/source/WMTS';
import { create as createTransform, compose as composeTransform, } from 'ol/transform';
import { getCenter as getExtentCenter } from 'ol/extent';
import { transform2D } from 'ol/geom/flat/transform';
const getAbsoluteUrl_ = (url) => {
    const a = document.createElement('a');
    a.href = encodeURI(url);
    return decodeURI(a.href);
};
const scratchOpacityCanvas = document.createElement('canvas');
export function asOpacity(canvas, opacity) {
    const ctx = scratchOpacityCanvas.getContext('2d');
    scratchOpacityCanvas.width = canvas.width;
    scratchOpacityCanvas.height = canvas.height;
    ctx.globalAlpha = opacity;
    ctx.drawImage(canvas, 0, 0);
    return scratchOpacityCanvas;
}
/**
 * Return the WMTS URL to use in the print spec.
 */
const getWmtsUrl_ = (source) => {
    const urls = source.getUrls();
    console.assert(urls.length > 0);
    return getAbsoluteUrl_(urls[0]);
};
export default class MapfishPrintBaseEncoder {
    url;
    /**
     * Provides a function to create app.print.Service objects used to
     * interact with MapFish Print v3 services.
     *
     */
    constructor(printUrl) {
        this.url = printUrl;
    }
    async createSpec(map, scale, printResolution, dpi, layout, format, customAttributes, customizer) {
        const mapSpec = await this.encodeMap(map, scale, printResolution, dpi, customizer);
        const attributes = {
            map: mapSpec,
        };
        Object.assign(attributes, customAttributes);
        return {
            attributes,
            format,
            layout,
        };
    }
    async getStatus(ref) {
        return await (await fetch(`${this.url}/status/${ref}.json`)).json();
    }
    async requestReport(spec) {
        const report = await fetch(`${this.url}/report.${spec.format}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(spec),
        });
        return await report.json();
    }
    // FIXME: add timeout
    // FIXME: handle errors
    getDownloadUrl(response, interval = 1000) {
        return new Promise((resolve, reject) => {
            const intervalId = setInterval(async () => {
                const status = await this.getStatus(response.ref);
                if (status.done) {
                    clearInterval(intervalId);
                    resolve(`${this.url}/report/${response.ref}`);
                }
            }, interval);
        });
    }
    async mapToLayers(map, printResolution, customizer) {
        const mapLayerGroup = map.getLayerGroup();
        console.assert(!!mapLayerGroup);
        const layerStates = mapLayerGroup
            .getLayerStatesArray()
            .filter(customizer.layerFilter)
            .sort((state, nextState) => (state.zIndex || 0) - (nextState.zIndex || 0))
            .reverse();
        const layers = [];
        for (const layerState of layerStates) {
            console.assert(printResolution !== undefined);
            const spec = await this.encodeLayer(layerState, printResolution, customizer);
            if (spec) {
                if (Array.isArray(spec)) {
                    layers.push(...spec);
                }
                else {
                    layers.push(spec);
                }
            }
        }
        return layers;
    }
    encodeTileLayer(layerState, customizer) {
        const layer = layerState.layer;
        console.assert(layer instanceof olLayerTile);
        const source = layer.getSource();
        if (source instanceof olSourceWMTS) {
            return this.encodeTileWmtsLayer(layerState, customizer);
        }
        else {
            return null;
        }
    }
    encodeTileWmtsLayer(layerState, customizer) {
        const layer = layerState.layer;
        console.assert(layer instanceof olLayerTile);
        const source = layer.getSource();
        console.assert(source instanceof olSourceWMTS);
        const dimensionParams = source.getDimensions();
        const dimensions = Object.keys(dimensionParams);
        // FIXME: remove "as const"
        const wmtsLayer = {
            type: 'wmts',
            baseURL: getWmtsUrl_(source),
            dimensions,
            dimensionParams,
            imageFormat: source.getFormat(),
            name: layer.get('name'),
            layer: source.getLayer(),
            matrices: getWmtsMatrices(source),
            matrixSet: source.getMatrixSet(),
            opacity: layerState.opacity,
            requestEncoding: source.getRequestEncoding(),
            style: source.getStyle(),
            version: source.getVersion(),
        };
        customizer.wmtsLayer(layerState, wmtsLayer, source);
        return wmtsLayer;
    }
    drawFeaturesToContext(features, styleFunction, resolution, coordinateToPixelTransform, vectorContext, additionalDraw) {
        if (!styleFunction) {
            return;
        }
        features.forEach((f) => {
            const optGeometry = f.getGeometry();
            if (!optGeometry) {
                return;
            }
            const geometry = optGeometry.clone();
            geometry.applyTransform((flatCoordinates, dest, stride) => {
                return transform2D(flatCoordinates, 0, flatCoordinates.length, stride || 2, coordinateToPixelTransform, dest);
            });
            const styles = styleFunction(f, resolution);
            if (styles) {
                if (Array.isArray(styles)) {
                    styles.forEach((style) => {
                        vectorContext.setStyle(style);
                        vectorContext.drawGeometry(geometry);
                    });
                }
                else {
                    vectorContext.setStyle(styles);
                    vectorContext.drawGeometry(geometry);
                }
                additionalDraw(geometry);
            }
        });
    }
    createCoordinateToPixelTransform(printExtent, resolution, size) {
        const coordinateToPixelTransform = createTransform();
        const center = getExtentCenter(printExtent);
        // See VectorImageLayer
        // this.coordinateToVectorPixelTransform_ = compose(this.coordinateToVectorPixelTransform_,
        composeTransform(coordinateToPixelTransform, size[0] / 2, size[1] / 2, 1 / resolution, -1 / resolution, 0, -center[0], -center[1]);
        return coordinateToPixelTransform;
    }
}
