// @geoblocks/mapfishprint/lib/utils.js
import WMTSTileGrid from 'ol/tilegrid/WMTS.js';
import { toSize } from 'ol/size.js';
import { Constants, CalculatedConstants } from "./constants.js";
/**
 * @param mapPageSize The page size in pixels (width, height)
 * @param center The coordinate of the extent's center.
 * @param scale The scale to calculate the extent width.
 * @returns an extent that fit the page size. Calculated with DPI_PER_DISTANCE_UNIT (by default using meters)
 */
export function getPrintExtent(mapPageSize, center, scale) {
    const [mapPageWidthMeters, mapPageHeightMeters] = mapPageSize.map((side) => ((side / CalculatedConstants.DPI_PER_DISTANCE_UNIT()) * scale) / 2);
    return [
        center[0] - mapPageWidthMeters,
        center[1] - mapPageHeightMeters,
        center[0] + mapPageWidthMeters,
        center[1] + mapPageHeightMeters,
    ];
}
/**
 * Takes a hex value and prepends a zero if it's a single digit.
 *
 * @param hex Hex value to prepend if single digit.
 * @returns hex value prepended with zero if it was single digit,
 *     otherwise the same value that was passed in.
 */
export function colorZeroPadding(hex) {
    return hex.length === 1 ? `0${hex}` : hex;
}
/**
 * Converts a color from RGB to hex representation.
 *
 * @param rgb rgb representation of the color.
 * @returns hex representation of the color.
 */
export function rgbArrayToHex(rgb) {
    const r = rgb[0];
    const g = rgb[1];
    const b = rgb[2];
    if (r !== (r & 255) || g !== (g & 255) || b !== (b & 255)) {
        throw new Error(`"(${r},${g},${b})" is not a valid RGB color`);
    }
    const hexR = colorZeroPadding(r.toString(16));
    const hexG = colorZeroPadding(g.toString(16));
    const hexB = colorZeroPadding(b.toString(16));
    return `#${hexR}${hexG}${hexB}`;
}
export function getWmtsMatrices(source) {
    const projection = source.getProjection();
    const tileGrid = source.getTileGrid();
    console.assert(tileGrid instanceof WMTSTileGrid);
    const matrixIds = tileGrid.getMatrixIds();
    const wmtsMatrices = [];
    const metersPerUnit = projection.getMetersPerUnit();
    console.assert(!!metersPerUnit);
    for (let i = 0; i < matrixIds.length; i++) {
        const tileRange = tileGrid.getFullTileRange(i);
        const resolutionMeters = tileGrid.getResolution(i) * metersPerUnit;
        wmtsMatrices.push({
            identifier: matrixIds[i],
            scaleDenominator: resolutionMeters / Constants.WMTS_PIXEL_SIZE,
            tileSize: toSize(tileGrid.getTileSize(i)),
            topLeftCorner: tileGrid.getOrigin(i),
            matrixSize: [tileRange.maxX - tileRange.minX + 1, tileRange.maxY - tileRange.minY + 1],
        });
    }
    return wmtsMatrices;
}
const scratchOpacityCanvas = document.createElement('canvas');
export function asOpacity(canvas, opacity) {
    const ctx = scratchOpacityCanvas.getContext('2d');
    scratchOpacityCanvas.width = canvas.width;
    scratchOpacityCanvas.height = canvas.height;
    ctx.globalAlpha = opacity;
    ctx.drawImage(canvas, 0, 0);
    return scratchOpacityCanvas;
}
export function getAbsoluteUrl(url) {
    const a = document.createElement('a');
    a.href = encodeURI(url);
    return decodeURI(a.href);
}
/**
 * Return the WMTS URL to use in the print spec.
 */
export function getWmtsUrl(source) {
    const urls = source.getUrls();
    console.assert(urls.length > 0);
    return getAbsoluteUrl(urls[0]);
}
export async function getStatus(mfpBaseUrl, ref) {
    const response = await fetch(`${mfpBaseUrl}/status/${ref}.json`);
    return await response.json();
}
export async function cancelPrint(mfpBaseUrl, ref) {
    const response = await fetch(`${mfpBaseUrl}/cancel/${ref}`, { method: 'DELETE' });
    return { status: response.status };
}
export async function requestReport(mfpBaseUrl, spec) {
    const report = await fetch(`${mfpBaseUrl}/report.${spec.format}`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(spec),
    });
    return await report.json();
}
/**
 * @param requestReport the name of the requested report
 * @param response The initial print response.
 * @param interval (s) the internal to poll the download url.
 * @param timeout (s) A timeout for this operation.
 * @returns a Promise with the download url once the document is printed or an error.
 */
export async function getDownloadUrl(requestReport, response, interval = 1000, timeout = 30000) {
    let totalDuration = 0 - interval;
    return new Promise((resolve, reject) => {
        const intervalId = setInterval(async () => {
            let status;
            try {
                status = await getStatus(requestReport, response.ref);
                if (status.error) {
                    throw new Error(status.error);
                }
            }
            catch (error) {
                reject(error);
            }
            if (status.done) {
                clearInterval(intervalId);
                resolve(`${requestReport}/report/${response.ref}`);
            }
            totalDuration += interval;
            if (totalDuration >= timeout) {
                clearInterval(intervalId);
                reject(new Error('Print duration exceeded'));
            }
        }, interval);
    });
}
